inc/functions/files.php

```diff
@@ -818,15 +818,34 @@ function rocket_clean_domain( $lang = '' ) {
 	$urls = apply_filters( 'rocket_clean_domain_urls', $urls, $lang );
 	$urls = array_filter( $urls );
 
+	$cache_path = rocket_get_constant( 'WP_ROCKET_CACHE_PATH' );
+
+	try {
+		$cache = new RecursiveDirectoryIterator( $cache_path, FilesystemIterator::SKIP_DOTS );
+	} catch ( UnexpectedValueException $e ) {
+		// No logging yet.
+		return;
+	}
+
+	try {
+		$iterator = new RecursiveIteratorIterator( $cache, RecursiveIteratorIterator::CHILD_FIRST );
+	} catch ( Exception $e ) {
+		// No logging yet.
+		return;
+	}
```

Comment:
@Tabrisrp This code is changing the behavior of what files/directories are passed to `rocket_rrmdir()` and the resultant deletion. 

With `glob()`:

1. It's passing the matching files/dirs from the given root, but not the child files/dirs within each level. It's not iterative.
2. It leaves the root directory(ies), i.e. deletes what is inside of each.

## What is this code doing?

It's recursively getting _all_ files/directories within the given cache and matching root pattern. **ALL** It's changing the behavior of this function and impacting performance when we run `rocket_rrmdir()`.

## How so?

Because it's recursively grabbing all levels of files/directories, it runs `rocket_rrmdir()` for each of those entries. 

## Why is that problematic? 

Let's look at what's happening in `rocket_rrmdir()`.

 Each time it's called:

- What's passed to it:
    - `glob()`: the entries inside of the root, but not the root itself.
   - this PR's code: the root + entries inside of it
         - Result: the root is also deleted.
- It fires the action `'before_rocket_rrmdir'`. 
     - With this code: same as `glob()` plus the number of the root files/directories. Result: Fired more times than `glob()`.
  - If it's a file, it's deleted and bails out.
  - If it's a directory:
       - it runs `glob` to get a list of all child entries at that level.
       - it filters the list to remove those that are set to "preserve"
       - And then it recursively iterates over the child entries returned from its `glob`.
       - So it's scanning the filesystem of each given directory. 
           - That's fine with the current behavior. 
           - With this new code, we're _rescanning the filesystem for each directory_. "Rescanning" is not performant.
  - For each directory, it fires `after_rocket_rrmdir`. <- directories only. 
      - Works the same for both `glob` and this new code.

Okay. So the code introduced in this PR has a negative impact on `rocket_rrmdir()` in terms of memory and processing speed.

## What can we do to preserve the current behavior, but get rid of `glob()`?

Replace both `try/catch` with:

```php
try {
	$iterator = new FilesystemIterator( $cache_path, FilesystemIterator::SKIP_DOTS );
} catch ( Exception $e ) {
	// No logging yet.
	return;
}
```

Reply:
You're right, thank you for catching that. I'm updating the code.

For later, do you think it would be more efficient to not rely on `rocket_rrmdir` for this function, and instead grab everything we need with the iterators, instead of the multiple `glob` calls?

> It leaves the root directory(ies), i.e. deletes what is inside of each.

This part is kind of an oversight from the original code in fact, the directories for each cache path _should_ be deleted when all their content is deleted too. But we can work on that later too.

Reply:
Done in https://github.com/wp-media/wp-rocket/pull/2510/commits/186cae0ddae49e17ecc2f3bb050003b07542015a

Reply:
>For later, do you think it would be more efficient to not rely on `rocket_rrmdir` for this function, and instead grab everything we need with the iterators, instead of the multiple glob calls?

Yes, absolutely. When we refactor the filesystem functions into our own `Filesystem` implementation, then we can look at reducing the redundancy and optimizing everything. 

For example, these `try/catch` chunks of code will be held within the new `Filesystem` class. That'll reduce the code where we use it.